[dashboard] New, list view (react)

nytai opened the pull request:
### CATEGORY

Choose one

- [ ] Bug Fix
- [ ] Enhancement (new features, refinement)
- [x] Refactor
- [ ] Add tests
- [ ] Build / Development Environment
- [ ] Documentation

### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
Replaces the current dashboard list view generated by FAB with one that is client rendered in react using the FAB/dashboard api. Design has tried to stay as close to the current design as possible, with some minor improvements. There have been recent efforts to redesign most of the views in superset, so having pages that are rendered in react will greatly facilitate those changes. The work in this PR is intended to lay the foundation for all list views that are to come. 

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
<!--- Skip this if not applicable -->
#### old view
<img width="1163" alt="Screen Shot 2019-12-20 at 9 34 47 AM" src="https://user-images.githubusercontent.com/10255196/71278296-05140800-230c-11ea-92b1-6f4e6cb2e844.png">

#### new view
<img width="1676" alt="Screen Shot 2019-12-20 at 9 30 06 AM" src="https://user-images.githubusercontent.com/10255196/71278105-828b4880-230b-11ea-9c4c-da88bc54f4f5.png">
<img width="1146" alt="Screen Shot 2019-12-20 at 9 30 54 AM" src="https://user-images.githubusercontent.com/10255196/71278144-9df65380-230b-11ea-9f9a-95bd309ecf11.png">
<img width="1149" alt="Screen Shot 2019-12-20 at 9 29 56 AM" src="https://user-images.githubusercontent.com/10255196/71278180-b5cdd780-230b-11ea-9705-10ebe078a0d5.png">


### TEST PLAN
<!--- What steps should be taken to verify the changes -->
- The welcome page works as before.
- The dashboards page renders, sorts, and filters
### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [ ] Has associated issue:
- [x] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### REVIEWERS

nytai committed:
adds dashboard listview component

nytai committed:
use new api

nytai committed:
use json over rison

nytai committed:
lint

nytai committed:
adds seperate dashboard list view

nytai committed:
edit and delete actions

nytai committed:
fix lint ignore

nytai committed:
fix common_bootstrap_payload is now a function

nytai committed:
fix license

nytai committed:
fix pylint

nytai committed:
isort

nytai committed:
fix tests

dpgaspar commented on superset/assets/src/dashboardList/DashboardList.tsx:
```diff
@@ -0,0 +1,243 @@
+            initialSort={this.initialSort}
+            filterable
+            filterTypes={[
+              { label: 'Starts With', value: 'sw' },
```
This object can be fetched from the API and you get translations for free also. This is on the `_info` endpoint

nytai committed:
lint

nytai committed:
lint ts

nytai committed:
fix js tests

dpgaspar commented on superset/assets/src/dashboardList/DashboardList.tsx:
```diff
@@ -0,0 +1,252 @@
+    },
+    {
+      Cell: ({ row: { state, original } }: any) => {
+        const handleDelete = () => this.handleDashboardDeleteConfirm(original);
```
On the MVC implementation if the user does not have permission to delete, the button is not rendered

The `_info` endpoint provides the user permissions on the class permission context

aboganas commented:
This is awesome, I hope this be ported to Flask-AppBuilder lists

nytai committed:
fix double import from bad rebase

nytai committed:
fix indent error

nytai committed:
lookup permissions

nytai committed:
generic permission lookup

dpgaspar commented on superset/models/dashboard.py:
```diff
@@ -179,6 +179,33 @@ def dashboard_link(self) -> Markup:
+
+    @property
+    def json(self):
+        return self.changed_by.to_json()
```
The output from this includes an hashed password, you can selectively add related fields using dotted notation on `list_columns`, ex:

```
    list_columns = [
        "id",
        "dashboard_title",
        "url",
        "published",
        "owners_json",
        "changed_by.username",
        "changed_by_name",
        "changed_by_url",
        "changed_on",
    ]
```

nytai committed:
get tslint to pass

nytai committed:
adds js specs

nytai committed:
lint

nytai committed:
fix rebase

nytai commented on superset/assets/jest.config.js:
```diff
@@ -32,4 +32,11 @@ module.exports = {
+  globals: {
+    'ts-jest': {
+      diagnostics: {
+        warnOnly: true,
```
so type errors don't cause jest tests to fail (they're still reported though)

nytai commented on superset/assets/src/types/react-table.d.ts:
```diff
@@ -0,0 +1,224 @@
+// Type definitions for react-table 7
```
the newest version of `react-table` doesn't have types in https://github.com/DefinitelyTyped/DefinitelyTyped

found these online and adjusted to for the features we're using.

nytai commented on superset/assets/tslint.json:
```diff
@@ -1,10 +1,11 @@
+    "interface-name": [true, "never-prefix"],
+    "quotemark": [true, "single"],
+    "jsx-no-multiline-js": false,
+    "jsx-no-lambda": false
```
We don't have these rules enabled in eslint and I found them to be very tedious. Opened https://github.com/apache/incubator-superset/issues/8865 for larger discussion

nytai commented on superset/views/dashboard/views.py:
```diff
@@ -102,6 +105,21 @@ def new(self):  # pylint: disable=no-self-use
         db.session.commit()
         return redirect(f"/superset/dashboard/{new_dashboard.id}/?edit=true")
 
+    @has_access
```
found I had to run `superset init` after adding this decorator. Not sure if there's a way around it. This is necessary to override FAB's default list view.

dpgaspar replied:
One possible way could be to override with:
```
Class Dashboard(BaseSupersetView):
     class_permission_name = "DashboardModelView"
         method_permission_name = {
        "new": "add",
        "list_all": "list",
    }
```

Note: This would deprecate the Dashboard.new permission

nytai committed:
lint

nytai committed:
lint again

nytai committed:
fix type errors preventing build

nytai committed:
adds more specs

nytai committed:
fix tslint error

nytai committed:
fix null check

nytai commented:
@etr2460 @graceguo-supercat @michellethomas Would be great to get some 👀 on this from folks who spend more of their time in the frontend.

dpgaspar commented on superset/models/dashboard.py:
```diff
@@ -179,6 +179,10 @@ def dashboard_link(self) -> Markup:
+    @property
+    def dashboard_title_raw(self) -> str:
+        return self.dashboard_title
+
```
Is this used?

nytai replied:
nope, thanks for flagging!

dpgaspar commented on superset/views/dashboard/api.py:
```diff
@@ -172,7 +172,7 @@ class DashboardRestApi(DashboardMixin, BaseSupersetModelRestApi):
         "related": "list",
     }
-    exclude_route_methods = ("info",)
+    # exclude_route_methods = ("info",)
```
remove?

nytai replied:
done

dpgaspar reviewed (changes requested):
Tested it locally and looks good, but noticed the following:
- when switching languages the user is redirected to the old dashboard list (previous MVC page)
- column labels are not translated
- Button labels are not translated
- Edit form back button does not work work properly (previous MVC), I would say that it's only usable when using modal forms form add and edit

nytai committed:
remove unecessary code

nytai committed:
use translations provided by api

nytai committed:
more translations

nytai commented:
@dpgaspar I've updated with the changes your requested as best I could. 

>when switching languages the user is redirected to the old dashboard list (previous MVC page)

looks like the user is redirected to the edit page of a specific dashboard, not sure how that's happening but doubt it's due to any change I made. Seems like that url is exposed in FAB and I don't really understand how that works right now. 

>column labels are not translated

I've updated the code to use the `label_columns` data provided by the backend. Looks like we're missing some translations there. 

>Button labels are not translated

I've updated the code to use the `t('....')` function from superset-ui/translation which seems to be the way of handling translations on the frontend. We're also likely missing most translations there. 

>Edit form back button does not work work properly (previous MVC), I would say that it's only usable when using modal forms form add and edit

This is true. looks like that links to `/back` which appears to be an endpoint in FAB? Somehow it's redirecting back to the charts list. Not sure how that works or how it's configured. -- There's work being done to move that functionality to a modal so that could be the workaround solution for that issue.

nytai committed:
linting

nytai committed:
fix spec

mistercrunch commented on superset/assets/src/views/dashboardList/DashboardList.tsx:
```diff
@@ -0,0 +1,294 @@
+
+const PAGE_SIZE = 25;
+
+interface Props {
```
I'm unclear about the pros/cons of using `interface`  over `propTypes`, assuming this is more typescript-friendly (?). Tradeoff seems to be consistency vs being more ts friendly (?)

nytai replied:
We could use both. PropTypes are checked at runtime by react in development and log to the console. Interface/types are checked by the typescript compiler/vscode and produce errors at build time and are highlighted in the editor. I'd say typescript is favorable once the entire repo converts to typescript as the types will be available in the editor (on hover, cmd click, etc). For now I'm happy to also add PropTypes until we transition more to typescript. 

FWIW: The react dev team recommends flow/typescript over PropTypes for larger projects: https://reactjs.org/docs/static-type-checking.html

mistercrunch commented on superset/assets/src/views/dashboardList/DashboardList.tsx:
```diff
@@ -0,0 +1,294 @@
+        <Modal show={this.state.showDeleteModal} onHide={this.toggleModal}>
+          <Modal.Header closeButton={true} />
+          <Modal.Body>
+            Are you sure you want to delete{' '}
```
`t()`

mistercrunch commented on superset/assets/src/welcome/DashboardTable.jsx:
```diff
@@ -18,34 +18,106 @@
+    search: PropTypes.string,
+  };
+
+  state = {
```
Is this the same as setting `state` in the constructor?

nytai replied:
Yes, they are essentially the same thing. The class properties is just sugar for setting variables in constructor. I personally prefer this over using the constructor as the constructor requires a call to `super()` which can sometimes be overlooked.

mistercrunch commented on superset/assets/src/views/dashboardList/DashboardList.tsx:
```diff
@@ -0,0 +1,294 @@
+    return this.hasPerm('can_delete');
+  }
+
+  public static propTypes = {
```
Wondered why we haven't used this syntax before and found this:
<img width="739" alt="Screen Shot 2020-01-14 at 11 52 03 AM" src="https://user-images.githubusercontent.com/487433/72377404-5671a600-36c4-11ea-903e-1c5e019c0e96.png">

nytai replied:
I've added the babel plugin for `transform-class-properties`.  The proposal for these is currently in stage 3, I believe, which is the step right before formal adoption into the standard. The proposal states that implementation is already under way in the latest chrome and in node 12. 
proposal: 
https://github.com/tc39/proposal-class-fields#class-field-declarations-for-javascript

Happy to switch back too. FWIW, every react project I've worked on has had this syntax enabled by babel.

nytai committed:
i18n

nytai committed:
fix register order
